pkg/resources/server.go

```diff
@@ -0,0 +1,282 @@
+	return &pluginapi.PreStartContainerResponse{}, nil
+}
+
+func (rs *resourceServer) GetDevicePluginOptions(ctx context.Context, empty *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	return &pluginapi.DevicePluginOptions{
+		PreStartRequired: false,
+	}, nil
+}
+
+func (rs *resourceServer) Init() error {
+	resourceName := rs.resourcPool.GetResourceName()
+	glog.Infof("initializing %s device pool", resourceName)
+	if err := rs.resourcPool.DiscoverDevices(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// gRPC server related
+func (rs *resourceServer) Start() error {
+	resourceName := rs.resourcPool.GetResourceName()
+	_ = rs.cleanUp() // try tp clean up and continue
+	glog.Infof("starting %s device plugin endpoint at: %s\n", resourceName, rs.endPoint)
+	sockPath := filepath.Join(types.SockDir, rs.endPoint)
+	lis, err := net.Listen("unix", sockPath)
+	if err != nil {
+		glog.Errorf("error starting %s device plugin endpoint: %v", resourceName, err)
+		return err
+	}
+
+	// Register all services
+	pluginapi.RegisterDevicePluginServer(rs.grpcServer, rs)
+
+	go rs.grpcServer.Serve(lis)
+	// Wait for server to start by launching a blocking connection
+	conn, err := grpc.Dial(sockPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithTimeout(5*time.Second),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	if err != nil {
+		glog.Errorf("error. unable to establish test connection with %s gRPC server: %v", resourceName, err)
+		return err
+	}
+	glog.Infof("%s device plugin endpoint started serving", resourceName)
+	conn.Close()
+
+	rs.triggerUpdate()
+
+	// Register with Kubelet.
+	err = rs.register()
+	if err != nil {
+		// Stop server
+		rs.grpcServer.Stop()
+		glog.Fatal(err)
+		return err
+	}
+	return nil
+}
+
+func (rs *resourceServer) restart() error {
+	resourceName := rs.resourcPool.GetResourceName()
+	glog.Infof("restarting %s device plugin server...", resourceName)
+	if rs.grpcServer == nil {
+		return fmt.Errorf("grpc server instance not found for %s", resourceName)
+	}
+	rs.grpcServer.Stop()
+	rs.grpcServer = nil
+	// Send terminate signal to ListAndWatch()
+	rs.termSignal <- true
+
+	rs.grpcServer = grpc.NewServer() // new instance of a grpc server
+	return rs.Start()
+}
+
+func (rs *resourceServer) Stop() error {
+	resourceName := rs.resourcPool.GetResourceName()
+	glog.Infof("stopping %s device plugin server...", resourceName)
+	if rs.grpcServer == nil {
+		return nil
+	}
+	// Send terminate signal to ListAndWatch()
+	rs.termSignal <- true
+	rs.stopWatcher <- true
+
+	rs.grpcServer.Stop()
+	rs.grpcServer = nil
+
+	return rs.cleanUp()
+}
+
+func (rs *resourceServer) Watch() {
+	// Watch for socket file; if not present restart server
+	sockPath := filepath.Join(types.SockDir, rs.endPoint)
+	for {
+		select {
+		case stop := <-rs.stopWatcher:
+			if stop {
+				return
+			}
+		default:
+			_, err := os.Lstat(sockPath)
```

Comment:
Definitely! This will reduce the burden of watching kubelet. Let's see if we can merge your kubelet plugin watcher PR with this implementation. This was implemented long before Kubelet plugin watcher support.